### PR TITLE
Eliminate XStream warnings originating from GeoServer

### PR DESCRIPTION
**Pull request by aaime**
There are quite a bit originating from GeoWebCache.

I've basically tried to follow the indication given by the XStream developers, and about to provide some feedback on the experience here: https://github.com/x-stream/xstream/issues/101

### CODE REVIEW

File: src/main/src/main/java/org/geoserver/config/util/SecureXStream.java [hunk 1]
```diff
@@ -158,6 +203,275 @@ protected MapperWrapper wrapMapper(MapperWrapper next) {
...
 
+    @Override
+    protected void setupConverters() {
+        Mapper mapper = getMapper();
+        ReflectionProvider reflectionProvider = getReflectionProvider();
+        ClassLoaderReference classLoaderReference = getClassLoaderReference();
+        ConverterLookup converterLookup = getConverterLookup();
+
+        registerConverter(new ReflectionConverter(mapper, reflectionProvider), PRIORITY_VERY_LOW);
+        registerConverter(new InternalBlackList(), PRIORITY_LOW);
```
Comment (tbarsballe) [inline src/main/src/main/java/org/geoserver/config/util/SecureXStream.java]: If this is a blacklist, should it really be low priority? Shouldn't it be one of the first converters to check.

File: src/main/src/main/java/org/geoserver/config/util/XStreamPersister.java [hunk 2]
```diff
@@ -658,8 +658,8 @@ void log(Level level, String msg) {
      */
     public void save(Object obj, OutputStream out) throws IOException {
         // unwrap dynamic proxies
-        obj = unwrapProxies(obj);
-        xs.toXML(obj, new OutputStreamWriter(out, "UTF-8"));
+        Object unwrapped = unwrapProxies(obj);
+        xs.toXML(unwrapped, new OutputStreamWriter(out, "UTF-8"));
```
Comment (tbarsballe) [inline src/main/src/main/java/org/geoserver/config/util/XStreamPersister.java]: This seems functionally the same, any reason for the change? (I don't especially care either way)

File: src/wfs/src/test/java/org/geoserver/wfs/GetFeatureTest.java [hunk 3]
```diff
@@ -459,6 +459,7 @@ public void testCQLFilter() throws Exception {
 
         String request = "wfs?request=GetFeature&typename=" + layer + "&version=1.0.0&service=wfs";
         Document doc = getAsDOM(request);
+        print(doc);
```
Comment (tbarsballe) [inline src/wfs/src/test/java/org/geoserver/wfs/GetFeatureTest.java]: Does this need to be here?

File: src/main/src/main/java/org/geoserver/config/util/SecureXStream.java [hunk 4]
```diff
@@ -224,4 +538,27 @@ public ForbiddenClassExceptionEx(String message, Throwable cause) {
             super(message, cause);
         }
     }
+
+    private class InternalBlackList implements Converter {
```
[the remark this answers is not in the document]
Comment (aaime) [reply src/main/src/main/java/org/geoserver/config/util/SecureXStream.java]: It's a verbatim copy from the XStream class

File: src/main/src/main/java/org/geoserver/config/util/TreeMapConverter.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+import com.thoughtworks.xstream.mapper.Mapper;
+import java.util.Comparator;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * Converts a java.util.TreeMap to XML, and serializes the associated java.util.Comparator. The
+ * converter assumes that the entries in the XML are already sorted according the comparator.
+ *
+ * <p>Cloned from XStream in order to avoid illegal reflective lookup warnings
```
[the remark this answers is not in the document]
Comment (aaime) [reply src/main/src/main/java/org/geoserver/config/util/TreeMapConverter.java]: Will fix

File: src/main/src/main/java/org/geoserver/config/util/TreeSetConverter.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import com.thoughtworks.xstream.mapper.Mapper;
+import java.util.Comparator;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+/**
+ * Converts a java.util.TreeSet to XML, and serializes the associated java.util.Comparator. The
+ * converter assumes that the elements in the XML are already sorted according the comparator.
+ *
+ * <p>Cloned from XStream in order to avoid illegal reflective lookup warnings
```
[the remark this answers is not in the document]
Comment (aaime) [reply src/main/src/main/java/org/geoserver/config/util/TreeSetConverter.java]: Will fix

File: src/main/src/main/java/org/geoserver/config/util/XStreamPersister.java (hunk 2 above)
Comment (aaime) [reply src/main/src/main/java/org/geoserver/config/util/XStreamPersister.java]: Debugging with changing objects made me nuts, was hard to compare before and after

File: src/wfs/src/test/java/org/geoserver/wfs/GetFeatureTest.java (hunk 3 above)
Comment (aaime) [reply src/wfs/src/test/java/org/geoserver/wfs/GetFeatureTest.java]: Nope, leftover, good catch

File: src/main/src/main/java/org/geoserver/config/util/SecureXStream.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (aaime) [reply src/main/src/main/java/org/geoserver/config/util/SecureXStream.java]: However, I could add a comment stating where it comes from

[the remark this answers is not in the document]
Comment (tbarsballe) [reply src/main/src/main/java/org/geoserver/config/util/SecureXStream.java]: A comment with the origin of that class would be great.

File: src/main/src/main/java/org/geoserver/config/util/XStreamPersister.java (hunk 2 above)
Comment (tbarsballe) [reply src/main/src/main/java/org/geoserver/config/util/XStreamPersister.java]: Thats a good reason, fair enough.

File: src/main/src/main/java/org/geoserver/config/util/SecureXStream.java (hunk 1 above)
Comment (aaime) [reply src/main/src/main/java/org/geoserver/config/util/SecureXStream.java]: Again, vertatim copy from XStream (the whole method is a copy of the XStream default one, with a minimum amount of adaptations)

Comment (tbarsballe) [reply src/main/src/main/java/org/geoserver/config/util/SecureXStream.java]: Okay, I withdraw my concerns then (still seems wrong, but I'm content enough for it to be an XStream problem and not a GeoServer one)

File: src/main/src/main/java/org/geoserver/config/util/TreeMapConverter.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (aaime) [reply src/main/src/main/java/org/geoserver/config/util/TreeMapConverter.java]: Turns out that I cannot... I add the </p> and the google formatter removes it!

[the remark this answers is not in the document]
Comment (tbarsballe) [reply src/main/src/main/java/org/geoserver/config/util/TreeMapConverter.java]: @jodygarnett - Have you run into this when doing your other javadoc changes? Did you find a workaround?

[the remark this answers is not in the document]
Comment (tbarsballe) [reply src/main/src/main/java/org/geoserver/config/util/TreeMapConverter.java]: Well, I guess leave it for now and we can revisit if it actually gives errors on release/publish for GeoServer?

**Comment by aaime** [comment]
@tbarsballe see the comment at https://github.com/x-stream/xstream/issues/101#issuecomment-432755816
Basically, XStream won't be modified, we can get rid of some of the bits that I've added, and hope that the treemap/treeset are not going to cause troubles for us (they should not, we have very simple usages of them in the configuration).

File: src/main/src/main/java/org/geoserver/config/util/TreeMapConverter.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jodygarnett) [reply src/main/src/main/java/org/geoserver/config/util/TreeMapConverter.java]: Well they figured out how to turn off the javadoc linter

**Pull request merged**